src/linear_algebra/matrix/circulant.lean

```diff
@@ -0,0 +1,240 @@
+/-
+Copyright (c) 2021 Lu-Ming Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lu-Ming Zhang
+-/
+import linear_algebra.matrix.symmetric
+import data.polynomial.monomial
+import data.matrix.pequiv
+import data.equiv.fin
+
+/-!
+# Circulant matrices
+
+This file contains the definition and basic results about circulant matrices.
+Given a vector `v : I → α` indexed by a type that is endowed with subtraction,
+`circulant v` is the matrix whose `(i,j)`th entry is `v (i = j)`.
+
+## Main results
+
+- `matrix.circulant`: the circulant matrix generated by a given vector `v : I → α`.
+
+## Implementation notes
+
+`fin.foo` is the `fin n` version of `foo`.
+Namely, the index type of the circulant matrices in discussion is `fin n`.
+
+## Tags
+
+circulant, matrix
+-/
+
+variables {α β I J R : Type*} {n : ℕ}
+
+namespace matrix
+
+open function
+open_locale matrix big_operators
+
+/-- Given the condition `[has_sub I]` and a vector `v : I → α`,
+    we define `circulant v` to be the circulant matrix generated by `v` of type `matrix I I α`.
+    The `(i,j)`th entry is defined to be `v (i - j)`. -/
+def circulant [has_sub I] (v : I → α) : matrix I I α
+| i j := v (i - j)
+
+lemma circulant_col_zero_eq [add_group I] (v : I → α) :
+  (λ i, (circulant v) i 0) = v :=
+by ext; simp [circulant]
+
+lemma circulant_injective [add_group I] : injective (circulant : (I → α) → matrix I I α) :=
+begin
+  intros v w h,
+  rw [← circulant_col_zero_eq v, ← circulant_col_zero_eq w, h]
+end
+
+lemma fin.circulant_injective : ∀ n, injective (λ v : fin n → α, circulant v)
+| 0     := dec_trivial
+| (n+1) := circulant_injective
+
+@[simp] lemma circulant_inj [add_group I] {v w : I → α} :
+  circulant v = circulant w ↔ v = w :=
+circulant_injective.eq_iff
+
+@[simp] lemma fin.circulant_inj {v w : fin n → α} :
+  circulant v = circulant w ↔ v = w :=
+(fin.circulant_injective n).eq_iff
+
+lemma transpose_circulant [add_group I] (v : I → α) :
+  (circulant v)ᵀ =  circulant (λ i, v (-i)) :=
+by ext; simp [circulant]
+
+lemma conj_transpose_circulant [has_star α] [add_group I] (v : I → α) :
+  (circulant v)ᴴ = circulant (star (λ i, v (-i))) :=
+by ext; simp [circulant]
+
+lemma fin.transpose_circulant : ∀ {n} (v : fin n → α), (circulant v)ᵀ =  circulant (λ i, v (-i))
+| 0     := dec_trivial
+| (n+1) := transpose_circulant
+
+lemma fin.conj_transpose_circulant [has_star α] :
+  ∀ {n} (v : fin n → α), (circulant v)ᴴ = circulant (star (λ i, v (-i)))
+| 0     := dec_trivial
+| (n+1) := conj_transpose_circulant
+
+lemma map_circulant [has_sub I] (v : I → α) (f : α → β) :
+  (circulant v).map f = circulant (λ i, f (v i)) :=
+by ext; simp [circulant]
```

Comment:
I would guess you can replace all the proofs like this with
```
ext $ λ _ _, rfl
```

Reply:
I don't think `ext $ λ _ _, rfl` will do for the most lemmas. Perhaps some virant.